### PR TITLE
fix(web): obtain media device label

### PR DESCRIPTION
**Pull request by WinterJack002**
Fix the issue where Firefox browser device list cannot be read.

### CODE REVIEW

**Comment by WinterJack002** [comment]
Fix the issue where Firefox browser device list cannot be read, replace 'label' with 'kind+DeviceId'.

File: webapp/components/device.tsx [hunk 1]
```diff
@@ -11,8 +11,47 @@ import SvgAudio from './svg/audio'
 import SvgVideo from './svg/video'
 import { SvgPresentCancel, SvgPresentToAll } from './svg/present'
 
-export default function DeviceBar(props: { streamId: string }) {
-  const [permissionAudio, setPermissionAudio] = useState("")
+// 1.将每个设备的 deviceId 和 label 转换为可显示的格式。如果 label 为空，会使用设备类型和 deviceId 作为显示名称。
```
Comment (a-wing) [inline webapp/components/device.tsx:14]: Use English comment. please

File: webapp/components/device.tsx [hunk 2]
```diff
@@ -11,8 +11,47 @@ import SvgAudio from './svg/audio'
...
+function deviceInfoToOption(info: MediaDeviceInfo) {
+  const value = info.deviceId;
+  let text = info.label;  
+  if (text.length <= 0) {
+      text = `${info.kind} (${info.deviceId})`;
+  }
+  return { value, text };
+}
+
+// 2.uniqByValue 去重
```
Comment (a-wing) [inline webapp/components/device.tsx:24]: English comment

File: webapp/components/device.tsx [hunk 3]
```diff
@@ -48,7 +87,7 @@ export default function DeviceBar(props: { streamId: string }) {
       // NOTE:
       // Chrome: audio_capture, video_capture
       // Safari: microphone, camera
-      if (status.name === "audio_capture" || "microphone") {
+      if (status.name === "audio_capture" || "microphone") { // "microphone"作为string不一直为ture吗？
```
Comment (a-wing) [inline webapp/components/device.tsx:90]: `(status.name === "audio_capture" || status.name === "microphone")`

File: webapp/components/device.tsx [hunk 4]
```diff
@@ -58,21 +97,58 @@ export default function DeviceBar(props: { streamId: string }) {
 
   const updateDeviceList = async () => {
     const devices = (await navigator.mediaDevices.enumerateDevices()).filter(i => !!i.deviceId)
-    const audios: Device[] = devices.filter(i => i.kind === 'audioinput')
-    const videos: Device[] = devices.filter(i => i.kind === 'videoinput')
+
+    // 使用 deviceInfoToOption加载设备列表
+    const audios = devices.filter(i => i.kind === 'audioinput').map(deviceInfoToOption)
+
+    console.log('this is audios:',audios)
```
Comment (a-wing) [inline webapp/components/device.tsx:104]: Remove this line

File: webapp/components/device.tsx [hunk 5]
```diff
@@ -58,21 +97,58 @@ export default function DeviceBar(props: { streamId: string }) {
...
+
+    const videos = devices.filter(i => i.kind === 'videoinput').map(deviceInfoToOption);
+    //const audios: Device[] = devices.filter(i => i.kind === 'audioinput')
+    //const videos: Device[] = devices.filter(i => i.kind === 'videoinput')
```
Comment (a-wing) [inline webapp/components/device.tsx:108]: Remove

File: webapp/components/device.tsx [hunk 6]
```diff
@@ -58,21 +97,58 @@ export default function DeviceBar(props: { streamId: string }) {
...
+
+    // 使用 uniqByValue 去重
+  const uniqueAudios = uniqByValue(audios);
+  console.log('this is uniqueAudios:',uniqueAudios)
```
Comment (a-wing) [inline webapp/components/device.tsx:112]: Remove this line

File: webapp/components/device.tsx [hunk 7]
```diff
@@ -58,21 +97,58 @@ export default function DeviceBar(props: { streamId: string }) {
...
-      if (device) await setCurrentDeviceVideo(device.deviceId)
+      let device = uniqueVideos[0]
+      if (device) {
+        try {
+        await setCurrentDeviceVideo(device.value)
+        console.log('Video device set successfully')
+      } catch (error) {
+        console.error('Failed to set video device:', error)
+      }
+    }else{
```
Comment (a-wing) [inline webapp/components/device.tsx:139]: `} else {`

File: webapp/components/device.tsx [hunk 8]
```diff
@@ -113,7 +189,7 @@ export default function DeviceBar(props: { streamId: string }) {
     setLoadingScreen(true)
     await onChangedDeviceVideo(userStatus.screen ? deviceNone.deviceId : deviceScreen.deviceId)
     setLoadingScreen(false)
-  }
+  }
```
Comment (a-wing) [inline webapp/components/device.tsx:192]: Trim tail ` `

File: webapp/components/device.tsx [hunk 9]
```diff
@@ -58,21 +97,58 @@ export default function DeviceBar(props: { streamId: string }) {
...
+      if (device) 
+        {
+          try {
+            await setCurrentDeviceAudio(device.value)
+          console.log('Audio device set successfully')
+        } catch (error) {
+          console.error('Failed to set audio device:', error)
+        }
+      }
+        //await setCurrentDeviceAudio(device.value)
```
Comment (a-wing) [inline webapp/components/device.tsx:127]: Remove unless line

File: webapp/components/device.tsx [hunk 10]
```diff
@@ -58,21 +97,58 @@ export default function DeviceBar(props: { streamId: string }) {
...
     }
 
     if (currentDeviceVideo === deviceNone.deviceId) {
-      let device = videos[0]
-      if (device) await setCurrentDeviceVideo(device.deviceId)
+      let device = uniqueVideos[0]
+      if (device) {
+        try {
+        await setCurrentDeviceVideo(device.value)
+        console.log('Video device set successfully')
```
Comment (a-wing) [inline webapp/components/device.tsx:135]:
```javascript
        try {
          await setCurrentDeviceVideo(device.value)
```

**Review by a-wing — CHANGES_REQUESTED** [review]
1. Use English comment. please
2. Remove unless line
3. Remove unless and debug `console.log`
4. Code style

* * *
About code style:

http://www.ruanyifeng.com/blog/2012/04/javascript_programming_style.html
https://www.cnblogs.com/Yellow-ice/p/15127392.html

**Comment by a-wing** [comment]
BTW: CLA. please. https://github.com/binbat/live777/blob/main/.github/CLA.md#list-of-contributors

You can read this example: https://github.com/binbat/live777/pull/95

**Comment by WinterJack002** [comment]
I do have configured the PGP key and signed it. However, the email address in my Git configuration is not the one for GPG signing (bound to GitHub). I just changed it with` git config --global user.email`. The next time I submit, it should show that it is verified.

File: webapp/components/device.tsx [hunk 11]
```diff
@@ -11,7 +11,40 @@ import SvgAudio from './svg/audio'
...
+    deviceId: item.value,         // use value as the device ID
+    kind,
+    label: item.text,             // use text as the label
+    groupId: '',
+
+  })as MediaDeviceInfo);
+}
+
 export default function DeviceBar(props: { streamId: string }) {
+
```
Comment (a-wing) [inline webapp/components/device.tsx:47]: This blank line is unused

File: webapp/components/device.tsx [hunk 12]
```diff
@@ -48,31 +81,55 @@ export default function DeviceBar(props: { streamId: string }) {
...
+    const uniqueAudios = uniqByValue(audios);
+    const uniqueVideos = uniqByValue(videos);
 
     if (currentDeviceAudio === deviceNone.deviceId) {
-      let device = audios[0]
-      if (device) await setCurrentDeviceAudio(device.deviceId)
+      let device = uniqueAudios[0]
+      if (device)
+        {
```
Comment (a-wing) [inline webapp/components/device.tsx:103]: `if (device) {`

File: webapp/components/device.tsx [hunk 13]
```diff
@@ -48,31 +81,55 @@ export default function DeviceBar(props: { streamId: string }) {
...
+          try {
+            await setCurrentDeviceAudio(device.value)
+          console.log('Audio device set successfully')
```
Comment (a-wing) [inline webapp/components/device.tsx:106]: Remove this log

File: webapp/components/device.tsx [hunk 14]
```diff
@@ -48,31 +81,55 @@ export default function DeviceBar(props: { streamId: string }) {
...
     }
 
     if (currentDeviceVideo === deviceNone.deviceId) {
-      let device = videos[0]
-      if (device) await setCurrentDeviceVideo(device.deviceId)
+      let device = uniqueVideos[0]
+      if (device) {
+        try {
+          await setCurrentDeviceVideo(device.value)
+        console.log('Video device set successfully')
```
Comment (a-wing) [inline webapp/components/device.tsx:119]: Remove this successfully log

File: webapp/components/device.tsx [hunk 15]
```diff
@@ -48,31 +81,55 @@ export default function DeviceBar(props: { streamId: string }) {
...
+      } catch (error) {
+        console.error('Failed to set video device:', error)
+      }
+    } else {
+      console.log('no video devices:')
+      await setCurrentDeviceVideo(deviceNone.deviceId)
+
+    }
```
Comment (a-wing) [inline webapp/components/device.tsx:127]: This block indent style error

File: webapp/components/device.tsx [hunk 16]
```diff
@@ -48,31 +81,55 @@ export default function DeviceBar(props: { streamId: string }) {
...
     }
 
-    setDeviceAudio([...audios])
-    setDeviceVideo([...videos, deviceScreen])
+    setDeviceAudio(convertToDevice(uniqueAudios,'audioinput'))
+    setDeviceVideo(convertToDevice(uniqueVideos,'videoinput'))
```
Comment (a-wing) [inline webapp/components/device.tsx:131]: `convertToDevice(uniqueVideos, 'videoinput')`

`,` after add a ` `

**Review by a-wing — CHANGES_REQUESTED** [review]
But, this patch has some problem is no resolve

This is use this patch is firefox:

Screenshot 2024-10-26 at 08 33 14

* * *

But this is live777 debugger show device in firefox

Screenshot 2024-10-26 at 08 32 49

**Comment by a-wing** [comment]
https://github.com/binbat/live777/commit/1fa54d423b5b6f53c15c1dbca2613005bd79b3f2#diff-a79fe0a1bc4d37db7cc13373925f80b8c81b4bb2ec609d32fb27d83d6c15c875R93-R96

**Comment by WinterJack002** [comment]
Correct the coding format and update issue solution.

**Pull request merged**